eth/catalyst/api.go

```diff
@@ -231,20 +307,31 @@ func (api *ConsensusAPI) ExecutePayloadV1(params beacon.ExecutableDataV1) (beaco
 	)
 	if td.Cmp(ttd) < 0 {
 		log.Warn("Ignoring pre-merge payload", "number", params.Number, "hash", params.BlockHash, "td", td, "ttd", ttd)
-		return api.invalid(), fmt.Errorf("cannot execute payload on top of pre-merge blocks: td %v, ttd %v", td, ttd)
+		return beacon.PayloadStatusV1{Status: beacon.INVALIDTERMINALBLOCK}, nil
+	}
+	if block.Time() <= parent.Time() {
+		log.Warn("Invalid timestamp", "parent", block.Time(), "block", block.Time())
+		return api.invalid(errors.New("invalid timestamp")), nil
 	}
 	log.Trace("Inserting block without sethead", "hash", block.Hash(), "number", block.Number)
 	if err := api.eth.BlockChain().InsertBlockWithoutSetHead(block); err != nil {
-		return api.invalid(), err
+		return api.invalid(err), nil
 	}
 	// We've accepted a valid payload from the beacon client. Mark the local
 	// chain transitions to notify other subsystems (e.g. downloader) of the
 	// behavioral change.
 	if merger := api.eth.Merger(); !merger.TDDReached() {
 		merger.ReachTTD()
 		api.eth.Downloader().Cancel()
+		// Shutdown gpu miner
+		if eng, ok := api.eth.Engine().(*beaconEngine.Beacon); ok {
+			if inner, ok := eng.InnerEngine().(*ethash.Ethash); ok {
+				inner.Close()
+			}
+		}
```

Comment:
Same as above, this feels very dangerous and we should definitely not do it in an API